### ISSUE TITLE
Define a format for errors and/or allow user defined error messages

### ISSUE THREAD

**Issue by Keats** [Keats/validator#10]
Right now it only returns the error code for the validation that failed, which is enough for me as I usually have control of both frontend and backend and just want to know which bit failed on which field.

However, having some more verbose error or user-defined ones (or both) would be welcome. 
For example when sending error message to a third party doing an API calls or if you don't control the frontend and need to send all the info (args used for the validator) to it.

A few things to consider:

### Serialize validators
Change the error type of `validate` from `Result<(), HashMap<String, Vec<String>>>` to `Result<(), HashMap<String, Vec<Validator>>>` and have `Validator` manually (so we can add their name) implement `Serialize` from Serde.

**Pros**
That's the easy option since this way we will get the name of the validator and all its args easily
**Cons**
Doesn't really work if we want custom error messages

### Custom message
Add an optional `error_msg` field to every single validator that will be sent with the error code. Sending only the error message makes it less flexible wrt if the frontend wants to change it, so we should it still have both error code and error. Error type will be `Result<(), HashMap<String, Vec<(String, Option<String>)>>>` 

### Default error messages
A neat way to do that would be to impl a function on Validator that returns a default error message. I would ignore i18n in default error messages as I don't think it will provide a good end user interface since end users will end up with errors like "Bank account number must match [0-9]{7}" or something like that that I got today with Transferwise. 
Proper error messages should be built by using the serialized validators in the frontend or just overwritten in the backend directly. There are also too many format for i18n and afaik not library for it in Rust yet.

The user could iterate on the errors hashmap or we could even impl a function on the hashmap directly so the user can do something like if they just want a hashmap with error messages:

```
match signup_data.validate() {
   Ok(()) => (),
   Err(e) => return e.get_error_messages()
};
```
which could be turned into a `validate_or_error_messages!` macro easily

If that approach works, it could be extended to transform errors into i18n objects potentially

## Current thoughts
I like the idea of serializing validators and having default error messages if wanted but I'm not sure where the custom error messages fit in that system. Ideally we'd have `String | Validator` as type for the value of hashmap but that's not possible in Rust. Another option is `(Validator, Option<String>)`.
The main issue would be what to do with the `custom` validator since you can have multiple kind of errors and how to work with serialization errors if we decide to let validator handle those (for consistent errors)

**Comment by theduke** [comment]
My proposal would be to add a new trait and common struct for validations.

I also need my validation errors to carry additional custom data, for i18n.

Imagine this error: "You must specify at least 3 regions, but you currently only entered 2".
The client contains formatting strings for each language, but it needs the following data, let's say in Json:
`errors: [{code: "missing_regions", "data": {"expected": 3, "actual": 2]}`

I see two options for the dynamic data. 
* Keep it generic across serialization formats and use https://github.com/dtolnay/erased-serde/, which provides a `Serialize` trait that is object safe. Then you can carry around a Box<Serialize> with the data.
* Just use `serde_json::Value` or https://github.com/arcnmx/serde-value, which can be cross-serialized to other formats easily if needed.

I think I'm in favor of using `serde-value::Value`.

All included validations would emit a SimpleValidationError with the error code as it is now, and with a standard message. (And maybe standard data as well, for example a length min validation would add as data: {min: 2, actual: 1}).

Both the code and the message can be overwritten like this:

```rust
  #[validate(len(min=3, code="my_code", msg="Get your shit together!"))]
  some_field: u64,
```

The structure I'm proposing also allows for nested struct validations.

```rust
// Error::description()  is for the error code.
// display() provides a nice error message in the default language
trait ValidationError : Error {
  fn data(&self) -> Option<&Value>
}

// All included validations emit a SimpleValidationError that 
struct SimpleValidationError {
  code: String,
  msg: String,
  data: Option<Value>,
}

impl fmt::Debug for SimpleValidationError {
  fn fmt(...) {
    write!(fmt, "{}" self.msg)
  }
}

impl Error for SimpleValidationError {
  fn description(&self) -> &str { &self.code }
  fn cause(&self) -> Option<&Error> { None }
}

struct FieldValidationError {
  field: String,
  // Consider just using SimpleValidationError here instead of Box<ValidationError>, since Simple.. should cover all use cases.
  errors: Vec<Box<ValidationError>>
}

struct MapValidationError {
  // Struct-level errors.
  errors: Vec<Box<ValidationError>>,
  // field errors.
  field_errors: Vec<FieldValidationError>,
}
```

**Comment by theduke** [comment]
Let me know what you think!

**Comment by Keats** [comment]
> I also need my validation errors to carry additional custom data, for i18n.

The tricky thing is if you want `validator` to return the error message ready for i18n directly.
For example `You must specify at least %d regions, but you currently only entered %d` and then you can have string translated using `gettext` or something similar directly but I guess you could have the message be the formatted string.
```rust
  #[validate(len(min=3, code="my_code", msg="Some field needs to have a length of at least 3, got %d"))]
  some_field: u64,
```

For your `data` type, I agree that a `HashMap<String, Value>` is probably the best bet.

The issue with giving a single error message on attribute is that sometimes you might want to have different messages depending on the error, but if you have the code and the values, you could build it yourself.

Overall it seems to make sense, I'll think about it for a couple of days and see how it feels then.

Thanks a lot for the input!

**Comment by lucab** [comment]
From a quick look at it, this discussion reminds me of the one going on at https://github.com/serde-rs/json/issues/199, so you may want to agree on a common approach to bubble up structured errors details there.

From my side I don't have any strong requirements, so something like `serde::de::Error` [trait](https://docs.serde.rs/serde/de/trait.Error.html) would be enough with its `custom()` method (but I have no idea about the whole i18n story).

**Comment by Leopard2A5** [comment]
I'm working on a PR for this. My approach is to not expose the actual validator (as i don't really need it in my use case) but to add error messages generated by the validator. For this I change the contents of the `Errors` struct from `HashMap<String, Vec<String>>` to `HashMap<String, Vec<Error>>` and Error is defined as

```rust
pub struct Error {
    validator: String,
    message: String,
}
```

I'll link the PR in here as soon as i think it's ready. Now I know it doesn't solve all the issues mentioned here I think it's a more or less big iteration on the way there.

**Comment by Keats** [comment]
@Leopard2A5 I would prefer going with the approach @theduke explained if possible
It allows i18n and is more flexible than an `Error` struct with just a `message` string where you don't get the arguments that were sent. Changing it to something else would cause unnecessary breaking changes imo

Sorry about the code in `validator_derive` too, it was very much a poc so it needs a good cleanup

**Comment by Leopard2A5** [comment]
@Keats I think we could use my PR as an iteration towards @theduke 's approach? I'm adding an Error struct already, we can extend it with the validator data in a subsequent step and add custom messages and things like that afterwards. What do you think?

Also: No worries, even though the macro is pretty long, the good indentation makes it relatively easy to work with. Weren't you unhappy about the constant repetition for the Optional stuff though? :laughing:

**Comment by Keats** [comment]
> I think we could use my PR as an iteration towards @theduke 's approach? I'm adding an Error struct already, we can extend it with the validator data in a subsequent step and add custom messages and things like that afterwards. What do you think? (Also, @theduke 's suggestion post was in february, idk if they lost interest?)

It's an intermediate step but not necessarily compatible with the step after (the `MapValidationError` struct, the traits etc). 
We can't expect people to have time to implement every idea, he/she already did a very good job with the design idea! Even I didn't work on this crate for ages since I am not currently using it.

> Weren't you unhappy about the constant repetition for the Optional stuff though? 😆
Yes I was.
Apparently the next version of `syn` will have a better UX (https://github.com/dtolnay/syn/issues/142) but I didn't read enough to see if it would affect this a lot or not.


I'm also curious how we can fit #22 after that

**Comment by theduke** [comment]
@Keats the reason for a trait was to have flexibility for the error structure and allow validators to return the appropriate error.

This is pretty much *neccessary* if you want to allow nested struct validation, since the errors for a struct look quite different from a regular field error. Check my `MapValidationError` from the code snippet above and think about a struct which contains another struct.

An idea  was also that certain validators could return custom errors that hold the error specific data, which can be converted to `HashMap<String, Value>` but be a more lightweight representation.

For examlpe, imagine a regexp validation error.

```rust
#[derive(Serialize, Deserialize)]
struct FormatValidationError {
  pub expected_format: String,
  pub actual_value: String,
}

impl ValidationError for FormatValidationError {
  fn data(&self) -> Value {
    return serte_json::value::to_value(self).unwrap()
  }
}
```
So I think a trait is neccessary.
An alternative to the `Box` solution would be an enum.

```rust
enum ValidationErr {
  Simple(SimpleValidationError),
  Map(MapValidationError),
  ...?,
  Custom(Box<ValidationError>),
}
```

**Comment by Keats** [comment]
I've started working on a new version (https://github.com/Keats/validator/tree/revamp) that uses structs for now. I'm refactoring the code as well so it will be much easier to change to a trait if it ends up being better. So for now that means an error is like https://github.com/Keats/validator/blob/revamp/validator/src/types.rs#L9-L13 (more on the `Value` later) and you can add a message or a code easily (and override a default one) for any validators. For your regex example that could be something like

```rust
    #[derive(Debug, Validate)]
    struct MyFile {
        // whatever error message
        #[validate(regex(path="MY_REGEX", code= "invalid-format", message = "The filename %(value) is valid. Only files of format xxx-01.js are allowed"))]
        filename: String,
    }
```

Each error will get the params of that validator: `value` for the value being validated (except for schema level validation where they don't get a `value` param) and the validator params when it makes sense (eg the path to a function or a regex are useless for a validation error). The user can also add whatever param in a custom function validation or a schema validation.

I'll finish up this approach soonish as I'm refactoring things along and changing to a trait later will be very easy if needed compared to the current master.

**Comment by theduke** [comment]
Great stuff, let me know if I can help out!

I personally don't see an issue with depending on serde.

The alternative would be implementing a custom `Value` type similar to serde_json::Value, and implement Serialize and deserialize for it hidden behind a feature.

**Comment by Keats** [comment]
First draft released as 0.6.0
